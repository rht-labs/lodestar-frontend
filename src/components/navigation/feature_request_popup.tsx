import React, { useState, useEffect } from 'react';
import { Popover, Button } from '@patternfly/react-core';

export const FeatureRequestPopup = ({
  children,
}: {
  children: React.ReactChild;
}) => {
  const featureRequestPopupKey = 'hasShownFeatureRequestPopup';

  const [
    hasShownFeatureRequestPopup,
    setHasShownFeatureRequestPopup,
  ] = useState<boolean>(true);

  useEffect(() => {
    const hasBeenShownStoredState =
      localStorage.getItem(featureRequestPopupKey) === 'true';
    if (hasShownFeatureRequestPopup !== hasBeenShownStoredState) {
      setHasShownFeatureRequestPopup(hasBeenShownStoredState);
    }
    if (!hasBeenShownStoredState) {
      setTimeout(() => {
        setHasShownFeatureRequestPopup(true);
      }, 5000);
    }
  }, [hasShownFeatureRequestPopup]);
  const onDismissed = () => {
    setHasShownFeatureRequestPopup(true);
    localStorage.setItem(featureRequestPopupKey, 'true');
  };
  return (
    <Popover
      isVisible={!hasShownFeatureRequestPopup}
      headerContent={<span>We're trying to improve</span>}
      shouldClose={onDismissed}
      distance={0}
      bodyContent={
        <div>
          <div>
            Tell us about a feature you would like to see included in Open
            Management Portal.
          </div>
          <Button onClick={onDismissed}>Ok, got it</Button>
        </div>
      }
    >
      {children as any}
    </Popover>
  );
};
