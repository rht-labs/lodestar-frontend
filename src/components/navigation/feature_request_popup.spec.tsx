import React from 'react';
import { render, act } from '@testing-library/react';
import { FeatureRequestPopup } from './feature_request_popup';

describe('Feature Request Popup', () => {
  test('renders properly', () => {
    // this code fixes an error setting up the test that is related to PopperJS
    // @see https://stackoverflow.com/questions/60333156/how-to-fix-typeerror-document-createrange-is-not-a-function-error-while-testi
    (global as any).document.createRange = () => ({
      setStart: () => {},
      setEnd: () => {},
      commonAncestorContainer: {
        nodeName: 'BODY',
        ownerDocument: document,
      },
    });
    //
    
    jest.useFakeTimers();
    act(() => {
      jest.runAllTimers();
    });
    expect(
      render(
        <div>
          <FeatureRequestPopup>
            <div />
          </FeatureRequestPopup>
        </div>
      )
    ).toMatchSnapshot();
  });
});
