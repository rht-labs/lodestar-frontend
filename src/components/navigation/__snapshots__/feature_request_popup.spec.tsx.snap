// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feature Request Popup renders properly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          aria-expanded="true"
        />
      </div>
    </div>
    <div
      class="tippy-popper"
      style="z-index: 9999; visibility: visible; transition-duration: 0ms; position: absolute;"
      x-placement="top"
    >
      <div
        class="tippy-tooltip pf-popover-theme pf-c-popover"
        data-animation="fade"
        data-arrow=""
        data-interactive=""
        data-state="hidden"
        id="tippy-1"
        role="tooltip"
        style="max-width: calc(2rem + 18.75rem); transition-duration: 0ms;"
        tabindex="-1"
      >
        <div
          class="tippy-arrow"
        />
        <div
          class="tippy-content"
          data-state="hidden"
          style="transition-duration: 0ms;"
        >
          <div>
            <div>
              <div
                aria-describedby="popover-pf-random-id-0-body"
                aria-labelledby="popover-pf-random-id-0-header"
                aria-modal="true"
                class=""
                role="dialog"
              >
                <div
                  class="pf-c-popover__content"
                >
                  <button
                    aria-label="Close"
                    class="pf-c-button pf-m-plain"
                    data-ouia-component-type="PF4/Button"
                    data-ouia-safe="true"
                    style="pointer-events: auto;"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      style="vertical-align: -0.125em;"
                      viewBox="0 0 352 512"
                      width="1em"
                    >
                      <path
                        d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                        transform=""
                      />
                    </svg>
                  </button>
                  <h6
                    class="pf-c-title pf-m-xl"
                    id="popover-pf-random-id-0-header"
                  >
                    <span>
                      We're trying to improve
                    </span>
                  </h6>
                  <div
                    class="pf-c-popover__body"
                    id="popover-pf-random-id-0-body"
                  >
                    <div>
                      <div>
                        Tell us about a feature you would like to see included in Open Management Portal.
                      </div>
                      <button
                        class="pf-c-button pf-m-primary"
                        data-ouia-component-type="PF4/Button"
                        data-ouia-safe="true"
                        type="button"
                      >
                        Ok, got it
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        aria-expanded="true"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
